### PR TITLE
Create rewriter for x86 strcpy chain

### PR DESCRIPTION
**Pull request by ptomin**
Many x86 binaries contanains strcpy(`<dst>`, `<src>`) compiled as
scasb/movsd/movsb sequence:
```
    	mov	edi,<src>
    	mov	edx,<dst>
    	or	ecx,0FFh
    	xor	eax,eax
    	repne scasb
    	not	ecx
    	sub	edi,ecx
    	mov	esi,edi
    	mov	eax,ecx
    	mov	edi,edx
    	shr	ecx,2h
    	rep movsd
    	mov	ecx,eax
    	and	ecx,3h
    	rep movsb
    	ret
```

### CODE REVIEW

File: src/Arch/X86/Analysis/StrcpyChainRewiter.cs [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+            if (intrinsic.Name != CommonOps.Strlen.Name)
+                return false;
+            if (!cmp.Equals(str, app.Arguments[0]))
+                return false;
+            if (bin.Right is not Constant c || !c.IsIntegerOne)
+                return false;
+            return true;
+        }
+
+        private Expression Expand(SsaState ssa, Expression e)
```
[the remark this answers is not in the document]
Comment (ptomin) [reply src/Arch/X86/Analysis/StrcpyChainRewiter.cs:210]: Logic: if id then defining expression else expression itself. Can it be named GetDefiningExpression?

[the remark this answers is not in the document]
Comment (uxmal) [reply src/Arch/X86/Analysis/StrcpyChainRewiter.cs:210]: `GetDefiningExpression` makes sense to me.

[the remark this answers is not in the document]
Comment (ptomin) [reply src/Arch/X86/Analysis/StrcpyChainRewiter.cs]: Done in 90a5a72

File: src/Arch/X86/Analysis/StrcpyChainRewiter.cs [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+            {
+                return false;
+            }
+            if (side.Expression is not Application app)
+                return false;
+            if (app.Procedure is not ProcedureConstant pc)
+                return false;
+            if (pc.Procedure is not IntrinsicProcedure intrinsic)
+                return false;
+            if (intrinsic.Name != CommonOps.Memcpy.Name)
```
[the remark this answers is not in the document]
Comment (ptomin) [reply src/Arch/X86/Analysis/StrcpyChainRewiter.cs]: Done in e7c52b8

File: src/Arch/X86/Analysis/StrcpyChainRewiter.cs [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+        public (SsaState, bool) Transform(SsaState subject)
+        {
+            bool changed = false;
+            foreach (var block in subject.Procedure.ControlGraph.Blocks)
+            {
+                var stms = block.Statements.ToArray();
+                for (int i = 0; i < stms.Length - 1; i++)
+                {
+                    if (TryRewriteStrcpy(subject, stms[i], stms[i + 1]))
```
[the remark this answers is not in the document]
Comment (ptomin) [reply src/Arch/X86/Analysis/StrcpyChainRewiter.cs:82]: This code excecutes after expression coalescing, so there is not statements between two `memcpy` calls normally. And it is unsafe to reorder statements in common case.

File: src/Arch/X86/Analysis/StrcpyChainRewiter.cs [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+                return e;
+            var defStm = ssa.Identifiers[id].DefStatement;
+            if (defStm is null)
+                return e;
+            if (defStm.Instruction is not Assignment ass)
+                return e;
+            return ass.Src;
+        }
+
+        private bool AsMemcpy(
```
[the remark this answers is not in the document]
Comment (ptomin) [reply src/Arch/X86/Analysis/StrcpyChainRewiter.cs:211]: Done in 1a90278.

File: src/Arch/X86/Analysis/StrcpyChainRewiter.cs [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+                return false;
+            if (!cmp.Equals(strSize1, strSize2))
+                return false;
+            var strSize = Expand(ssa, strSize1);
+            if (!IsStrlenInc1(strSize, str))
+                return false;
+            return true;
+        }
+
+        private bool IsAlignment4(
```
[the remark this answers is not in the document]
Comment (ptomin) [reply src/Arch/X86/Analysis/StrcpyChainRewiter.cs]: I haven't seen values other than 4. But it would be better to implement common case: see aef9603

**Comment by uxmal** [comment]
Looks good. Thanks for the contribution!

**Pull request merged**